### PR TITLE
CI: caching: closer match work/CI guarantees

### PR DESCRIPTION
**Pull request by Anton-Latukha**
* Would note we have a `cabal v2-build all --enable-tests --enable-benchmark` problem - that command requires the level of dependency consistency guarantees that PR CI does not checks.

`test` workflow does essentially `cabal v2-build --enable-tests` on matrix by target.
`bench` workflow does `cabal v2-build --enable-benchmarks` by target.

`caching` used `all --enable-tests --enable-benchmarks` - requiring total consistency.

These changes are to make `caching` able to work when `test` & `bench` deps
versions are inconsistent (which is how CI & people use them - separately).

* "`Workaround to CI platform`" needs to be removed either way, since I found out post hooks does not run if any of (even marked `continue-on-error: true`) steps fail.

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/2536"></a>

### CODE REVIEW

File: .github/workflows/caching.yml
Comment (jneira) [inline .github/workflows/caching.yml:185]: Not sure about this. I think we should optimize workflows for the "happy" path, the stable master branch and pr's not tweaking too much the project dependencies. 
`ghc-9.2.1` branch is somewhat a extraordinary one and as pepe commented there, optimize caching for it is not the main goal, but make it work. Authors of commits are maintainers testing changes locally or in their forks and fast ci is desirable but no essential imo.

**Comment by Anton-Latukha** [comment]
Note: "`Workaround to CI platform`" needs to be removed either way, because post hooks does not run if any of (even `continue-on-error: true`) steps fail.

File: .github/workflows/caching.yml
Comment (Anton-Latukha) [reply .github/workflows/caching.yml:185]: Well, yes this is also true.

This would've been a nice thing to have simply because, since we have manual runs inside PRs - this would be logical to have to cache through difficulties in the PR.

While `||` also ensures that after success commands on the right do not run, CI in `master` would terminate on first 4-5 of them - which would create a good cache. & that is why (because it is free to be added and skipped over) I was free to make that long line of commands "just in case" it runs inside PRs.

But at this point I also have no interest, that is a minor thing, this PR has what to attend to: *(updated the head post)*

**Comment by Anton-Latukha** [comment]
I reformulated the PR to address the most important stuff in it.

To match the process completely - we would need to run the `v2-build --enable-tests` command for every plugin separately, so it matches how `test` does it.

File: .github/workflows/caching.yml
Comment (Anton-Latukha) [reply .github/workflows/caching.yml]: But whould note, this was also the way to follow the provided by `test` & `bench` guarantees, while not requiring maintenance of https://github.com/haskell/haskell-language-server/pull/2536#issuecomment-1001551417

**Comment by Anton-Latukha** [comment]
[skip circleci]

File: .github/workflows/caching.yml
Comment (jneira) [inline .github/workflows/caching.yml:208]: Should not have this
```suggestion
        name: (For Bench workflow) Build benchmark targets
        continue-on-error: true
```
?

File: .github/workflows/caching.yml
Comment (jneira) [inline .github/workflows/caching.yml:199]: minor whitespace thing
```suggestion
          cabal $cabalBuild --enable-tests --only-download
```

File: .github/workflows/test.yml
Comment (jneira) [inline .github/workflows/test.yml:183]: not related with this pr but i've just was thinking if `cabal update` should not be before the freeze step, we have the hackage index fixed but freeze would change if the index change (only an observation, no asking for changes)

Comment (Anton-Latukha) [reply .github/workflows/test.yml:183]: `freeze` seems to save the real Hackage index used. So if `index-state` is pinned to something, the `freeze` infers the real snapshot time before that moment.

That's why it seems that the placement of `update` does not even matter really. It even probably may be removed.

But `update` would play an important role, if `haskell/actions/setup` for some reason has the older Hackage update than the `index-state` package uses. Currently *idk* if `haskell/actions/setup` does `cabal update`:

Screenshot-2021-12-27-17:06:32

so action reports nothing.

Currently, it does does the update: https://github.com/haskell/actions/blob/a64ac8cca0d4aec4e3fd4a5fbc9ae3c4953e20aa/setup/src/setup-haskell.ts#L75 (well it should update at every run).

**Review by jneira — APPROVED** [review]
lgtm, thanks!

**Comment by Anton-Latukha** [comment]
Nice review.

Would belabor additionally: now CI would start to cache `benchmarks` only to the same configuration that is used in `bench`. That, on update of GHC in `bench` to update `caching. But in that way we both deduplicate that build-cache over several workflows & cache only used `benchmarks` dependencies.

**Comment by Anton-Latukha** [comment]
After the recent merge `HLint` check works.

**Pull request merged**
